### PR TITLE
Challenge progress card

### PR DESCRIPTION
**Pull request by mohsinsurani**
Rebase completed

### CODE REVIEW

File: WalkingChallenge/Challenge/Challenge.swift [hunk 1]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+  private let lblTitle: UILabel = UILabel(typography: .headerTitle)
+  private let btnSettings: UIButton = UIButton(type: .system)
+  private let tblCards: UITableView = UITableView(frame: .zero)
+
+  private var arrCards: [JourneyCardMiles] = []
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    self.navigationController?.navigationBar.barTintColor = #colorLiteral(red: 0.9647058823, green: 0.9725490196, blue: 0.9803921568, alpha: 1.0000000000)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:43]: Is the color here not one of the standard colors from the AKF palette?  If so, why not?  If it is, why not use the color constants?

File: WalkingChallenge/Challenge/Challenge.swift [hunk 2]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+    self.navigationController?.navigationBar.setValue(true, forKey: "hidesShadow")
+
+    layout()
+  }
+
+  private func layout() {
+    view.backgroundColor = #colorLiteral(red: 0.9647058823, green: 0.9725490196, blue: 0.9803921568, alpha: 1.0000000000)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:50]: Similar

File: WalkingChallenge/Challenge/Challenge.swift [hunk 3]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+
+    lblTitle.textColor = .black
+
+    view.addSubview(lblTitle)
+    lblTitle.snp.makeConstraints {
+      $0.top.equalTo(topLayoutGuide.snp.bottom)
+      $0.left.right.equalToSuperview().inset(Style.Padding.p16)
+    }
+
+    btnSettings.setImage(UIImage(named: "GearIcon"), for: .normal)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:62]: IIRC there is a constant for the name here, please use that.

File: WalkingChallenge/Challenge/Challenge.swift [hunk 4]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+
+    view.addSubview(btnSettings)
+    btnSettings.snp.makeConstraints {
+      $0.top.equalTo(topLayoutGuide.snp.bottom)
+      $0.right.right.equalToSuperview().inset(Style.Padding.p16)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:67]: The button here will have the same constraints as the title label which seems wrong.

File: WalkingChallenge/Challenge/Challenge.swift [hunk 5]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+    }
+
+    tblCards.allowsSelection = false
+    tblCards.backgroundColor = #colorLiteral(red: 0.9647058823, green: 0.9725490196, blue: 0.9803921568, alpha: 1.0000000000)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:71]: Similar question about the color as above.

File: WalkingChallenge/Challenge/Challenge.swift [hunk 6]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+    tblCards.dataSource = self
+    tblCards.delegate = self
+    tblCards.separatorStyle = .none
+    tblCards.estimatedRowHeight = 44
+    tblCards.rowHeight = UITableView.automaticDimension
+
+
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:78]: Why the extra whitespace?

File: WalkingChallenge/Challenge/Challenge.swift [hunk 7]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+    tblCards.estimatedRowHeight = 44
+    tblCards.rowHeight = UITableView.automaticDimension
+
+    tblCards.register(JourneyMilesCardView.self)
+    tblCards.register(MilestoneDetailCardView.self)
+
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:84]: Why the whitespace?

File: WalkingChallenge/Challenge/Challenge.swift [hunk 8]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+      $0.bottom.equalToSuperview().inset(Style.Padding.p16)
+      $0.left.right.equalToSuperview().inset(Style.Padding.p16)
+    }
+
+    arrCards.append(JourneyCardMiles())
+
+    tblCards.reloadData()
+  }
+
+  private func getTitle(_ title: String, subTitle: String) -> NSMutableAttributedString {
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:99]: This seems unused, why was this added?

File: WalkingChallenge/Challenge/Challenge.swift [hunk 9]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+
+
+extension Challenge: UITableViewDelegate {
+  func tableView(_ tableView: UITableView,
+                 heightForHeaderInSection section: Int) -> CGFloat {
+    return Style.Padding.p16
+  }
+
+extension Challenge: UITableViewDelegate {
+//  func tableView(_ tableView: UITableView,
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:174]: Please delete the dead code.

File: WalkingChallenge/Challenge/Challenge.swift [hunk 10]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+
+    if indexPath.row == 0 {
+      let cell: UITableViewCell =
+        tableView.dequeueReusableCell(withIdentifier: card.renderer,
+                                      for: indexPath)
+      guard let view = cell as? JourneyMilesCardView else { return UITableViewCell() }
+      view.render(card)
+
+      view.showMileStone = {
+        self.arrCards[indexPath.section].expandMile(true)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:144]: Why is `expandMile` not a property?  If there is something that needs to occur on the set why not use an observable?  The `didSet` should allow you to post-process the change.

File: WalkingChallenge/Challenge/Challenge.swift [hunk 11]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+    } else {
+      let miles = card.miles[indexPath.row - 1]
+
+      let cell: UITableViewCell =
+        tableView.dequeueReusableCell(withIdentifier: miles.renderer,
+                                      for: indexPath)
+      guard let view = cell as? MilestoneDetailCardView else { return UITableViewCell() }
+      view.render(miles)
+
+      view.hideFirstLayer(indexPath.row == 1)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:158]: Again, an observable property is better I think.

File: WalkingChallenge/Challenge/Challenge.swift [hunk 12]
```diff
@@ -31,4 +31,161 @@ import UIKit
...
+      view.showLastRoundedCorner(indexPath.row == card.miles.count)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/Challenge.swift:159]: Similar

File: WalkingChallenge/Challenge/JourneyCard.swift [hunk 13]
```diff
@@ -0,0 +1,60 @@
+//
+//  JourneyCard.swift
+//  WalkingChallenge
+//
+//  Created by Mohsin on 24/02/19.
+//  Copyright © 2019 AKDN. All rights reserved.
+//
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCard.swift:7]: Please use the header as the rest of the project.

File: WalkingChallenge/Challenge/JourneyCard.swift [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+  required init?(coder aDecoder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+
+  private func initialise() {
+    guard let view = self as? StylizedJourneyView else { return }
+
+    self.contentView.backgroundColor = .white
+
+//    self.contentView.layer.cornerRadius = Style.Size.s8
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCard.swift:31]: Please remove the dead code.

File: WalkingChallenge/Challenge/JourneyCard.swift [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+//    self.layer.shadowRadius = Style.Size.s8
+
+    view.layout()
+  }
+}
+
+extension UIView {
+
+  func roundCorners(_ corners: CACornerMask, radius: CGFloat) {
+
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCard.swift:45]: Why the whitespace?

File: WalkingChallenge/Challenge/JourneyCard.swift [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    if #available(iOS 11.0, *) {
+      self.clipsToBounds = false
+      self.layer.cornerRadius = radius
+      self.layer.maskedCorners = corners// [.layerMinXMinYCorner, .layerMaxXMinYCorner]
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCard.swift:49]: Please remove the dead code.

File: WalkingChallenge/Challenge/JourneyCard.swift [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+//  Created by Mohsin on 24/02/19.
+//  Copyright © 2019 AKDN. All rights reserved.
+//
+
+import Foundation
+import SnapKit
+import Foundation
+import UIKit
+
+typealias StylizedJourneyView =  JourneyCardViewCell & LayoutableCardView
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCard.swift:14]: Why do you need a new type alias?  Can you not extend the `StylizedCarView` appropriately?

File: WalkingChallenge/Challenge/JourneyCardView.swift [hunk 18]
```diff
@@ -0,0 +1,79 @@
+//
+//  StylizedJourneyView.swift
+//  WalkingChallenge
+//
+//  Created by Mohsin on 24/02/19.
+//  Copyright © 2019 AKDN. All rights reserved.
+//
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCardView.swift:7]: Please use the standard header.

File: WalkingChallenge/Challenge/JourneyCardView.swift [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+  private var lblChallengeTimelineLabel: UILabel = UILabel(typography: .smallRegular)
+
+  func layout() {
+    addSubviews([imgImage, lblName, lblChallengeTimelineLabel])
+
+    guard let gear = UIImage(named: Assets.NotificationsSelected) else {
+      fatalError("missing asset \(Assets.NotificationsSelected)")
+    }
+
+    // TODO(compnerd) figure out the correct size for the image
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyCardView.swift:25]: Did you just copy/paste this code?

File: WalkingChallenge/Challenge/JourneyMilesCardView.swift [hunk 20]
```diff
@@ -0,0 +1,115 @@
+//
+//  JourneyMilesCardView.swift
+//  WalkingChallenge
+//
+//  Created by Mohsin on 10/05/19.
+//  Copyright © 2019 AKDN. All rights reserved.
+//
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyMilesCardView.swift:7]: Header

File: WalkingChallenge/Challenge/JourneyMilesCardView.swift [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+import Foundation
+
+class JourneyMilesCardView: StylizedJourneyView {
+  private var lblName: UILabel = UILabel(typography: .bodyBold)
+  private var lblChallengeTimelineLabel: UILabel = UILabel(typography: .smallRegular)
+  private var imgImage: UIImageView = UIImageView(frame: .zero)
+  private var lblWalkMiles: UILabel = UILabel(typography: .smallRegular)
+  private var btnViewMileStones: UIButton = UIButton(type: .system)
+
+  var showMileStone: (() -> Void)?
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyMilesCardView.swift:20]: Please give this an explicit accessibility.

File: WalkingChallenge/Challenge/JourneyMilesCardView.swift [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+    lblName.snp.makeConstraints {
+      $0.top.equalToSuperview().offset(Style.Padding.p16)
+      $0.left.equalToSuperview().inset(Style.Padding.p16)
+      $0.right.equalToSuperview().inset(Style.Padding.p16)
+    }
+
+    lblChallengeTimelineLabel.snp.makeConstraints {
+      $0.top.equalTo(lblName.snp.bottom).offset(Style.Padding.p8)
+      $0.left.equalTo(lblName.snp.left)
+      $0.right.equalToSuperview().inset(Style.Padding.p16)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyMilesCardView.swift:40]: Why not make this the same width as `lblName`?

File: WalkingChallenge/Challenge/JourneyMilesCardView.swift [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+      $0.top.equalTo(lblChallengeTimelineLabel.snp.bottom).inset(Style.Padding.p8)
+      $0.height.equalTo(szImageSize)
+      $0.width.equalTo(szImageSize)
+      $0.centerX.equalToSuperview()
+    }
+
+    lblWalkMiles.snp.makeConstraints {
+      $0.top.equalTo(imgImage.snp.bottom).offset(Style.Padding.p12)
+      $0.left.equalTo(lblName.snp.left)
+      $0.right.equalToSuperview().inset(Style.Padding.p8)
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyMilesCardView.swift:54]: Why not make constrain this to have the same right hand side as the `lblName`?

File: WalkingChallenge/Challenge/JourneyMilesCardView.swift [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+  }
+}
+
+struct JourneyCardMiles: Card {
+  let renderer: String = JourneyCardView.identifier
+
+  let title: String = "Journey"
+  let timeline: String = "1 out of 10 milestones completed"
+  let walkMiles: String = "365 miles left to walk untill your team's next milestone!"
+  let viewMileStone: String = "View Milestones"
```
Comment (compnerd) [inline WalkingChallenge/Challenge/JourneyMilesCardView.swift:108]: Can you move these into the Strings table?

File: WalkingChallenge/Challenge/MilestoneDetailCardView.swift [hunk 25]
```diff
@@ -0,0 +1,166 @@
+//
+//  MilestoneDetailCardView.swift
+//  WalkingChallenge
+//
+//  Created by Mohsin on 11/05/19.
+//  Copyright © 2019 AKDN. All rights reserved.
+//
```
Comment (compnerd) [inline WalkingChallenge/Challenge/MilestoneDetailCardView.swift:7]: Standard header please

File: WalkingChallenge/LocalizedStrings.swift.orig [hunk 26]
```diff
@@ -0,0 +1,175 @@
+/**
```
Comment (compnerd) [inline WalkingChallenge/LocalizedStrings.swift.orig:1]: I think that you added a stray file, please remove it

**Review by compnerd — CHANGES_REQUESTED** [review]
Please collapse the changes Into a single commit, this doesn't feel like 13 standalone changes.

**Comment by samisuteria** [comment]
Closing due to stale

**Pull request closed without merging**